DEVX-2184: "Detecting Abnormal Transactions" recipe migration (#604)
confluentinc/kafka-tutorials: _includes/tutorials/anomaly-detection/ksql/markup/dev/create-accounts-to-monitor-table.adoc

> A single transaction to one of the companies in the suspicious_names table is probably okay, but multiple transactions to one or more of those companies in a 24-hour period is alarming. ksqlDB gives us the ability to see if alarming activity is/was present for a particular user with the following query.

Reviewer: ```suggestion
For this use case, let's say that a single transaction to one of the companies in the suspicious_names table is probably okay, but multiple transactions to one or more of those companies in a 24-hour period is an anomaly. ksqlDB gives us the ability to see if any anomalies are present for a particular user with the following query.
```

[source,sql]
----
CREATE TABLE accounts_to_monitor
    WITH (kafka_topic='accounts_to_monitor', partitions=1, value_format='JSON') AS
    SELECT TIMESTAMPTOSTRING(WINDOWSTART, 'yyyy-MM-dd HH:mm:ss Z') AS WINDOW_START, // <1>
           TIMESTAMPTOSTRING(WINDOWEND, 'yyyy-MM-dd HH:mm:ss Z') AS WINDOW_END,
           USERNAME
    FROM suspicious_transactions
    WINDOW TUMBLING (SIZE 24 HOURS) // <2>
    GROUP BY USERNAME
    HAVING COUNT(*) > 3; // <3>
----
<1> The timestamps here are important, they tell us what interval of time suspicious activity occurred.

> <2> The `WINDOW TUMBLING` part of the query allows us to do an aggregation with distinct time boundaries. In this case our window is fixed at a length of 24 hours, does not allow gaps, and does not allow overlapping. Other types of windows are explained in the "Collect data over time" section of Kafka-Tutorials. If further explanation is need, checkout the link:https://docs.ksqldb.io/en/latest/concepts/time-and-windows-in-ksqldb-queries/#windows-in-sql-queries[*ksqlDB documentation about windows*]. It contains in depth descriptions and visualizations.

Reviewer: This is a lot of information that is essentially buried in a "footnote".  Readers often skip this as a result.  Can you move most of this out of the footnote and into the text area?

Author: This was a tricky request... The numbers in the code + "footnote" are called callouts in asciidoc and are meant for annotations. I did find a way to embed the callout icons in paragraphs... hopefully that is more apparent?

For this use case, let's say that a single transaction to one of the companies in the suspicious_names table is probably okay, but multiple transactions to one or more of those companies in a 24-hour period is an anomaly. ksqlDB gives us the ability to see if any anomalies are present for a particular user with the following query.

[source,sql]
----
CREATE TABLE accounts_to_monitor
    WITH (kafka_topic='accounts_to_monitor', partitions=1, value_format='JSON') AS
    SELECT TIMESTAMPTOSTRING(WINDOWSTART, 'yyyy-MM-dd HH:mm:ss Z') AS WINDOW_START, // <1>
           TIMESTAMPTOSTRING(WINDOWEND, 'yyyy-MM-dd HH:mm:ss Z') AS WINDOW_END,
           USERNAME
    FROM suspicious_transactions
    WINDOW TUMBLING (SIZE 24 HOURS) // <2>
    GROUP BY USERNAME
    HAVING COUNT(*) > 3; // <3>
----
+++++
<div class="colist arabic">

> <p><i class="conum" data-value="1"></i><b>1</b> The timestamps here are important, they tell us what interval of time suspicious activity occurred.</p>

Reviewer: Awesome presentation @awalther28

> <p><i class="conum" data-value="1"></i><b>1</b> The timestamps here are important, they tell us what interval of time suspicious activity occurred.</p>

Reviewer: `they tell us what interval of time suspicious activity occurred` -- the timestamps themselves don't tell the interval.  Would it be more accurate to phrase, something along the lines of, `they are used to calculate the interval of time between transactions` (or however you see fit)

Author: hm rereading this, I think I should have said "The fields `WINDOW_START` and `WINDOW_END`" rather than "timestamps". The number maps to the field creation of the windows. @ybyzek what are your thoughts on:
>The fields `WINDOW_START` and `WINDOW_END` tell us what interval of time suspicious activity occurred.

Reviewer: Actually, I misread as well! (I mistakenly was thinking of the timestamps of the transaction itself). Yes, your proposed text disambiguates.

Author: Change applied.

> <p><i class="conum" data-value="2"></i><b>2</b> The <code>WINDOW TUMBLING</code> part of the query allows us to do an aggregation with distinct time boundaries. In this case our window is fixed at a length of 24 hours, does not allow gaps, and does not allow overlapping. Other types of windows are explained in the "Collect data over time" section of Kafka-Tutorials. If further explanation is need, checkout the <a href="https://docs.ksqldb.io/en/latest/concepts/time-and-windows-in-ksqldb-queries/#windows-in-sql-queries"><strong>ksqlDB documentation about windows</strong></a>. It contains in depth descriptions and visualizations.</p>

Reviewer: ```suggestion
<p><i class="conum" data-value="2"></i><b>2</b> The <code>WINDOW TUMBLING</code> part of the query allows us to do an aggregation with distinct time boundaries. In this case our window is fixed at a length of 24 hours, does not allow gaps, and does not allow overlapping. Other types of windows are explained in the "Collect data over time" section of Kafka-Tutorials. For more in-depth descriptions and visualizations, checkout the <a href="https://docs.ksqldb.io/en/latest/concepts/time-and-windows-in-ksqldb-queries/#windows-in-sql-queries"><strong>ksqlDB documentation about windows</strong></a>. </p>
```

<p><i class="conum" data-value="3"></i><b>3</b> The last two lines of the query address how you would determine if a user had multiple suspicious transactions. This aspect of the query says, in essence, if any user has greater than 3 suspicious transactions during the window, emit an event to the accounts_to_monitor table. </p>

> </div>

Reviewer 2: Super minor nit: It wasn't obvious to me at first to create this table. Maybe add a sentence at the end, telling the user explicitly to create the table.  But only if you want to. Again this is a minor subjective comment.

Author: That's a great point, I did add a clause there.
